include/wlr/types/wlr_output.h

```diff
@@ -303,26 +294,90 @@ void wlr_output_lock_attach_render(struct wlr_output *output, bool lock);
+ * yourself.
+ *
+ * Hardware cursors are always in the ARGB8888 format.
  */
-void wlr_output_render_software_cursors(struct wlr_output *output,
-	pixman_region32_t *damage);
 
+/**
+ * Sets the size of the hardware cursor buffer to the values pointed to by
+ * x and y.
+ *
+ * Some backends have strict requirements for the cursor size and may not be
+ * able to use the size you requested exactly. The actual value used will be
+ * left in x and y. If this is not large enough for the cursor you wish to
+ * draw, you should fall back to using software cursors.
+ *
+ * This returns false if the backend does not support hardware cursors or from
+ * some other internal error. Either way, you should fall back to using
+ * software cursors.
+ */
+bool wlr_output_cursor_try_set_size(struct wlr_output *output, int *x, int *y);
+
+/**
+ * Attaches the renderer to the cursor buffer.
+ *
+ * You must successfully call `wlr_output_cursor_try_set_size` at least once
+ * before using this. Any rendering done will be applied on
+ * `wlr_output_cursor_commit`.
+ */
+bool wlr_output_cursor_attach_render(struct wlr_output *output, int *buffer_age);
 
-struct wlr_output_cursor *wlr_output_cursor_create(struct wlr_output *output);
 /**
- * Sets the cursor image. The image must be already scaled for the output.
+ * Enables displaying the hardware cursor.
+ *
+ * This does not affect the ability to render to the cursor. It only sets
+ * whether or not it will be displayed. By default, the cursor is disabled.
+ *
+ * This will be applied on `wlr_output_cursor_commit`.
+ */
+void wlr_output_cursor_enable(struct wlr_output *output, bool enable);
+
+/**
+ * Moves the position of the hardware cursor on the main framebuffer.
+ *
+ * This is in output buffer coordinates. Negative values and values that would
+ * place the cursor partially or entirely off the main framebuffer are allowed.
+ *
+ * The hotspot is the point on the cursor buffer (from the top left) where the
+ * cursor actually points. For example, with a cursor representing a hand, the
+ * hotspot may be over the index finger, rather than the top left corner.
+ *
+ * Some backends may not be able to move the cursor freely, but may still be
+ * able to make use of the hotspot to correctly show the cursor.
+ *
+ * This will be applied on `wlr_output_cursor_commit`.
+ */
+void wlr_output_cursor_move(struct wlr_output *output, int x, int y,
+	int hotspot_x, int hotspot_y);
+
+/**
+ * Apply pending cursor state.
+ *
+ * This may or may not be synchronized with the output, and would be applied on
+ * `wlr_output_commit`, depending on the backend limitations. However, it's
+ * advised to assume that they are sychronized and do any cursor operations at
+ * the same time you render to the output.
+ *
+ * If you have called `wlr_output_cursor_attach_render`, you must call this
+ * function before `wlr_output_attach_render`, otherwise you risk losing or
+ * corrupting rendering state.
```

Comment:
Can we make this an assertion instead of a "risk"?

Reply:
It becomes quite a pain to check reliably. Since the current EGL context/surface bound is global state, it could be bound to any output, even on a different backend.